overlord/devicestate/devicestate.go

```diff
@@ -1411,35 +1442,289 @@ func createRecoverySystemTasks(st *state.State, label string, snapSetupTasks []s
+	// set.
+	ValidationSets []*asserts.ValidationSet
+
+	// LocalSnaps is an optional list of snaps that will be used to create
+	// the new recovery system. If provided, this list must contain any snap
+	// that is not already installed that will be needed by the new recovery
+	// system.
+	LocalSnaps []LocalSnap
+
+	// TestSystem is set to true if the new recovery system should be verified
+	// by rebooting into the new system, prior to marking it as a valid recovery
+	// system. If false, the system will immediately be considered a valid
+	// recovery system.
+	TestSystem bool
+
+	// MarkCurrent is set to true if the new recovery system should be
+	// marked as the current system.
+	MarkCurrent bool
+}
+
+var ErrNoRecoverySystem = errors.New("recovery system does not exist")
+
+// CreateRecoverySystem creates a new recovery system with the given label. See
+// CreateRecoverySystemOptions for details on the options that can be provided.
+func CreateRecoverySystem(st *state.State, label string, opts CreateRecoverySystemOptions) (chg *state.Change, err error) {
 	var seeded bool
-	err := st.Get("seeded", &seeded)
+	err = st.Get("seeded", &seeded)
 	if err != nil && !errors.Is(err, state.ErrNoState) {
 		return nil, err
 	}
 	if !seeded {
 		return nil, fmt.Errorf("cannot create new recovery systems until fully seeded")
 	}
-	chg := st.NewChange("create-recovery-system", fmt.Sprintf("Create new recovery system with label %q", label))
-	ts, err := createRecoverySystemTasks(st, label, nil)
+
+	valsets := snapasserts.NewValidationSets()
+	for _, vs := range opts.ValidationSets {
+		if err := valsets.Add(vs); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := valsets.Conflict(); err != nil {
+		return nil, err
+	}
+
+	revisions, err := valsets.Revisions()
+	if err != nil {
+		return nil, err
+	}
+
+	model, err := findModel(st)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: this restriction should be lifted eventually (in the case that we
+	// have a dangerous model), and we should fall back to using snap names in
+	// places that IDs are used
+	if err := checkForSnapIDs(model, opts.LocalSnaps); err != nil {
+		return nil, err
+	}
+
+	// check that all snaps from the model are valid in the validation sets
+	if err := checkModelSnapsCanBePresentInValidationSets(model, valsets); err != nil {
+		return nil, err
+	}
+
+	// snaps required by validation sets must also be required by the model
+	if err := checkForRequiredSnapsNotRequiredInModel(model, valsets); err != nil {
+		return nil, err
+	}
+
+	tracker := snap.NewSelfContainedSetPrereqTracker()
+	offline := len(opts.LocalSnaps) > 0
+
+	var downloadTSS []*state.TaskSet
+	for _, sn := range model.AllSnaps() {
+		rev := revisions[sn.Name]
+
+		needsInstall, err := snapNeedsInstall(st, sn.Name, rev)
+		if err != nil {
+			return nil, err
+		}
+
+		if !needsInstall {
+			info, err := snapstate.CurrentInfo(st, sn.Name)
+			if err != nil {
+				return nil, err
+			}
+			tracker.Add(info)
```

Comment:
`tracker.Add(info)` is in the end called in all cases except when an error is returned, so maybe it could be called earlier in the loop to avoid the 3 calls in different parts.

Reply:
The `info` in all 3 cases is different, so it can't be lifted out into once place, unfortunately.

Reply:
Oh indeed, ignore me :slightly_smiling_face: